Reconstruction/RecCalorimeter/src/components/CreateCaloClustersSlidingWindow.cpp

```diff
@@ -0,0 +1,364 @@
+            << "Make sure you have GeoSvc and SimSvc in the right order in the configuration." << endmsg;
+    return StatusCode::FAILURE;
+  }
+  // check if readouts exist
+  if(m_geoSvc->lcdd()->readouts().find(m_readoutName) == m_geoSvc->lcdd()->readouts().end()) {
+    error()<<"Readout <<"<<m_readoutName<<">> does not exist."<<endmsg;
+    return StatusCode::FAILURE;
+  }
+  // retrieve PhiEta segmentation
+  m_segmentation = dynamic_cast<DD4hep::DDSegmentation::GridPhiEta*>(m_geoSvc->lcdd()->readout(m_readoutName).segmentation().segmentation());
+  if(m_segmentation == nullptr) {
+    error()<<"There is no phi-eta segmentation."<<endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  // get the ID of the volume for which the cells are counted
+  auto decoder = m_geoSvc->lcdd()->readout(m_readoutName).idSpec().decoder();
+  if(m_fieldNames.size() != m_fieldValues.size()) {
+    error() << "Size of names and values is not the same" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  for(uint it = 0; it < m_fieldNames.size(); it++) {
+    (*decoder)[m_fieldNames[it]] = m_fieldValues[it];
+  }
+  m_volumeId = decoder->getValue();
+  info() << "Reconstruction in volume with ID " << m_volumeId << endmsg;
+
+  info() << "CreateCaloClustersSlidingWindow initialized" << endmsg;
+  return StatusCode::SUCCESS;
+}
+
+StatusCode CreateCaloClustersSlidingWindow::execute() {
+  // 1. Get calorimeter towers (calorimeter grid in eta phi, all layers merged)
+  prepareTowers();
+  buildTowers();
+  debug()<<"Number of calorimeter towers (eta x phi) : "<<m_towers.size()<<" x "<<m_towers[0].size()<<endmsg;
+
+  // 2. Find local maxima with sliding window, build preclusters, calculate their barycentre position
+  // calculate the sum of first m_nEtaWindow bins in eta, for each phi tower
+  std::vector<float> sumOverEta (m_nPhiTower, 0);
+  for(int iEta = 0; iEta < m_nEtaWindow; iEta++) {
+    std::transform (sumOverEta.begin(), sumOverEta.end(), m_towers[iEta].begin(), sumOverEta.begin(), std::plus<float>());
+  }
+
+  // preclusters with phi, eta weighted position and transverse energy
+  std::vector<cluster> m_preClusters;
+  int halfEtaPos =  floor(m_nEtaPosition/2.);
+  int halfPhiPos =  floor(m_nPhiPosition/2.);
+  float posEta = 0;
+  float posPhi = 0;
+  float sumEnergyPos = 0;
+
+  // loop over all Eta slices starting at the half of the first window
+  int halfEtaWin =  floor(m_nEtaWindow/2.);
+  int halfPhiWin =  floor(m_nPhiWindow/2.);
+  float sumWindow = 0;
+  float sumPhiSlicePrevEtaWin = 0;
+  float sumPhiSliceNextEtaWin = 0;
+  float sumFirstPhiSlice = 0;
+  float sumLastPhiSlice = 0;
+  bool toRemove = false;
+  // one slice in eta = window, now only sum over window elements in phi
+  for(int iEta = halfEtaWin; iEta < m_nEtaTower - halfEtaWin; iEta++) {
+    // sum the first window in phi: for full phi coverage we start with phi bins from the end ...
+    for(int iPhiWindow = m_nPhiTower - halfPhiWin; iPhiWindow < m_nPhiTower; iPhiWindow++) {
```

Comment:
If I understand it correctly, for std::accumulate I specify range with ptrs, while for phi tower vector it can be a sum of elements from the end of it and from the beginning (last bin in phi is a direct neighbour of the first bin).